tools/dataflow-throttling/src/main/java/com/google/cloud/pso/dataflowthrottling/DynamicThrottlingTransform.java

```diff
@@ -0,0 +1,332 @@
+     * @param <InputT> Type of the event from the source.
+     */
+    public static class Result<InputT> implements Serializable {
+        private InputT input;
+        private String error;
+
+        public Result(InputT input, String error) {
+            this.input = input;
+            this.error = error;
+        }
+    }
+
+    /**
+     * This functionalInterface is used to pass the paylod to clientcall.
+     * @param <InputT> Type of the event from the source.
+     * @param <OutpuT> Type of the enriched event.
+     */
+    @FunctionalInterface
+    public interface ClientCall<InputT, OutpuT> extends Serializable {
+        OutpuT call(InputT n) throws Exception;
+    }
+
+    /**
+     * The {com.google.cloud.pso.dataflowthrottling.DynamicThrottlingTransform} constructor provides the custom execution options passed by the client.
+     * @param builder object which holds the user defined params.
+     */
+    public DynamicThrottlingTransform(Builder<InputT, OutputT> builder){
+        this.numberOfGroups= builder.numberOfGroups;
+        this.batchInterval= builder.batchInterval;
+        this.resetCounterInterval= builder.resetCounterInterval;
+        this.numOfEventsToBeProcessedForBatch= builder.numOfEventsToBeProcessedForBatch;
+        this.kInRejectionProbability= builder.kInRejectionProbability;
+        this.elemCoder= builder.elemCoder;
+        this.clientCall= builder.clientCall;
+    }
+
+    /**
+     * Collection of accepted requests response.
+     * @return PCollection<ClientCallResponses>
+     */
+    public TupleTag<OutputT> getSuccessTag(){
+        return successTag;
+    }
+
+    /**
+     * Collection of request responses of all error codes except 200,429.
+     * @return PCollection<ClientCall Responses Except 200and 429>
+     */
+    public TupleTag<Result<InputT>> getErrorTag(){
+        return errorTag;
+    }
+
+    /**
+     * Collection of rejected requests and Throttled requests.
+     * @return PCollection<Error_429 and ThrottledRequests>
+     */
+    public TupleTag<Result<InputT>> getThrottlingTag(){
+        return throttlingTag;
+    }
+
+    /**
+     * The Dynamic throttling transform begins here.
+     * @param input payload from the source.
+     * @return PCollectionTuple of tuple tags.
+     */
+    @Override
+    public PCollectionTuple expand(PCollection<InputT> input) {
+        /**
+         * {Grouping} It will segregate incoming events into single/multiple number of groups.
+         */
+        PCollection<KV<Integer, InputT>> groups=input.apply("Grouping", ParDo.of(new DoFn<InputT, KV<Integer, InputT>>() {
+            @DoFn.ProcessElement
+            public void processElement(ProcessContext context) {
+                context.output(KV.of(ThreadLocalRandom.current().nextInt(1, numberOfGroups + 1), context.element()));
+            }
+        }));
+
+        /**
+         * {Throttling}: Payload won't be processed from client-side based on the rejection probability.
```

Comment:
Payload is either sent to the external service or rejected based on the rejection probability.